### PR TITLE
Fix pickle vulnerability

### PR DESCRIPTION
**Pull request by deniskamazur**
This PR fixes a vulnerability in the hivemind server. For more information check out #381.

### CODE REVIEW

**Review by justheuristic — APPROVED** [review]
<:+1: />

It is strange, though, that the packb/unpackb is reported as not covered by tests

File: hivemind/utils/tensor_descr.py [hunk 1]
```diff
@@ -70,13 +83,55 @@ def from_tensor(cls, tensor: torch.Tensor, compression=CompressionType.NONE) ->
...
         return super().make_empty(size=(*batch_size, *self.shape[1:]), **kwargs)
 
+    def packb(self) -> bytes:
+        obj_dict = asdict(self)
+        obj_dict.pop("device")
+
+        if self.dtype is not None:
+            obj_dict["dtype"] = str(self.dtype)
+        else:
+            obj_dict.pop("dtype")
```
Comment (borzunov) [inline hivemind/utils/tensor_descr.py:100]:
```suggestion
```

I suggest to remove this (= keep `obj_dict["dtype"] = None` in this case) and replace `if "dtype" in obj_dict:` with the `if obj_dict["dtype"] is not None` line in `unpack()`.

If we do that for dtype, layout, etc., the code will become shorter and we will have a way to distinguish the None value from the absence of the field (may be useful if this class is extended in future).

**Review by borzunov — CHANGES_REQUESTED** [review]
Thanks for this important contribution!

Let's write tests with running `MSGPackSerializer.dumps/loads` for `BatchTensorDescriptor`. Also, please consider the minor comment above.

File: hivemind/utils/tensor_descr.py [hunk 2]
```diff
@@ -70,13 +83,48 @@ def from_tensor(cls, tensor: torch.Tensor, compression=CompressionType.NONE) ->
...
+            compression=compression if tensor.is_floating_point() else CompressionType.NONE,
         )
 
     def make_empty(self, *batch_size: int, **kwargs) -> torch.Tensor:
         assert self.shape[0] is None, "Make sure 0-th dimension is not specified (set to None)"
         return super().make_empty(size=(*batch_size, *self.shape[1:]), **kwargs)
 
+    def packb(self) -> bytes:
+        obj_dict = asdict(self)
+        obj_dict.pop("device")
```
Comment (borzunov) [inline hivemind/utils/tensor_descr.py:95]: Let's move this line right below `if self.device is not None:` (currently, reading the code surprised me with removing the device, and I got the reasoning only when I finished reading this function).

**Review by borzunov — APPROVED** [review]

File: hivemind/utils/tensor_descr.py [hunk 3]
```diff
@@ -70,13 +83,48 @@ def from_tensor(cls, tensor: torch.Tensor, compression=CompressionType.NONE) ->
...
+            obj_dict["dtype"] = _str_to_torch_type(obj_dict["dtype"], torch.dtype)
+
+        if obj_dict["layout"] is not None:
+            obj_dict["layout"] = _str_to_torch_type(obj_dict["layout"], torch.layout)
+
+        if obj_dict["device_type"] is not None:
+            obj_dict["device"] = torch.device(
+                obj_dict.pop("device_type"),
+                obj_dict.pop("device_index"),
+            )
```
Comment (borzunov) [inline hivemind/utils/tensor_descr.py:122]:
```suggestion
        if obj_dict["device_type"] is not None:
            obj_dict["device"] = torch.device(obj_dict["device_type"], obj_dict["device_index"])
        else:
            obj_dict["device"] = None
        del obj_dict["device_type"], obj_dict["device_index"]
```

**Pull request merged**